File: src/main/java/io/github/colintimbarndt/chat_emotes/config/ChatEmotesConfig.java
package io.github.colintimbarndt.chat_emotes.config;

import com.google.gson.*;
import com.google.gson.stream.JsonWriter;
import io.github.colintimbarndt.chat_emotes.ChatEmotesMod;
import io.github.colintimbarndt.chat_emotes.util.BomAwareReader;
import net.fabricmc.loader.api.Version;
import net.fabricmc.loader.api.VersionParsingException;
import net.minecraft.util.GsonHelper;
import org.jetbrains.annotations.NotNull;

import java.io.*;
import java.lang.reflect.Type;
import java.util.*;

public class ChatEmotesConfig {
    private static final Gson GSON = new GsonBuilder()
            .registerTypeAdapter(ResourcepackConfig.LIST_TYPE, new ResourcepackConfig.ListSerializer())
            .registerTypeAdapter(ChatEmotesConfig.class, new Serializer())
            .create();

    private final @NotNull List<ResourcepackConfig> resourcepacks;

    public ChatEmotesConfig() {
        this(Collections.emptyList());
    }

    private ChatEmotesConfig(@NotNull List<ResourcepackConfig> resourcepacks) {
        this.resourcepacks = resourcepacks;
    }

    public @NotNull List<ResourcepackConfig> getResourcepacks() {
        return resourcepacks;
    }

    public void save(OutputStream output) throws IOException {
        try (final var writer = new JsonWriter(new OutputStreamWriter(output))) {
            GSON.toJson(this, ChatEmotesConfig.class, writer);
        }
    }

    public static ChatEmotesConfig load(InputStream input) throws IOException, JsonParseException {
        try (final var reader = BomAwareReader.create(input)) {
            return GSON.fromJson(reader, ChatEmotesConfig.class);
        }
    }

    public static class Serializer implements JsonSerializer<ChatEmotesConfig>, JsonDeserializer<ChatEmotesConfig> {
        @Override
        public JsonElement serialize(ChatEmotesConfig config, Type type, JsonSerializationContext ctx) {
            final var root = new JsonObject();
            root.addProperty("version", ChatEmotesMod.getModMetadata().getVersion().getFriendlyString());
            root.add("resourcepacks", ctx.serialize(config.resourcepacks, ResourcepackConfig.LIST_TYPE));
            return root;
        }

        @Override
        public ChatEmotesConfig deserialize(
                JsonElement element,
                Type type,
                JsonDeserializationContext ctx
        ) throws JsonParseException {
            final var root = element.getAsJsonObject();
            if (!GsonHelper.isStringValue(root, "version")) {
                throw new JsonSyntaxException(
                        "Expected version to be a string, got " + GsonHelper.getType(root.get("version"))
                );
            }
            final Version version;
            try {
                version = Version.parse(GsonHelper.getAsString(root, "version"));
            } catch (VersionParsingException ex) {
                throw new JsonSyntaxException("Invalid version", ex);
            }
            if (version.compareTo(ChatEmotesMod.getModMetadata().getVersion()) != 0) {
                throw new JsonParseException("Unsupported version of config: " + version.getFriendlyString());
            }
            final List<ResourcepackConfig> resourcepacks = ctx.deserialize(
                    GsonHelper.getAsJsonObject(root, "resourcepacks"),
                    ResourcepackConfig.LIST_TYPE
            );
            return new ChatEmotesConfig(resourcepacks);
        }
    }
}

File: fabric_mod_server/src/main/kotlin/io/github/colintimbarndt/chat_emotes/config/ChatEmotesConfig.kt
package io.github.colintimbarndt.chat_emotes.config

import com.google.gson.*
import com.google.gson.stream.JsonWriter
import io.github.colintimbarndt.chat_emotes.ChatEmotesMod.Companion.modMetadata
import io.github.colintimbarndt.chat_emotes.config.ResourcepackConfig.ListSerializer
import io.github.colintimbarndt.chat_emotes.util.BomAwareReader.create
import net.fabricmc.loader.api.Version
import net.fabricmc.loader.api.VersionParsingException
import net.minecraft.util.GsonHelper
import java.io.*
import java.lang.reflect.Type

class ChatEmotesConfig private constructor(val resourcepacks: List<ResourcepackConfig>) {

    constructor() : this(emptyList<ResourcepackConfig>())

    @Throws(IOException::class)
    fun save(output: OutputStream) {
        JsonWriter(OutputStreamWriter(output)).use { writer -> GSON.toJson(this, ChatEmotesConfig::class.java, writer) }
    }

    object Serializer : JsonSerializer<ChatEmotesConfig>, JsonDeserializer<ChatEmotesConfig> {
        override fun serialize(config: ChatEmotesConfig, type: Type, ctx: JsonSerializationContext): JsonElement {
            val root = JsonObject()
            root.addProperty("version", modMetadata!!.version.friendlyString)
            root.add("resourcepacks", ctx.serialize(config.resourcepacks, ResourcepackConfig.LIST_TYPE))
            return root
        }

        @Throws(JsonParseException::class)
        override fun deserialize(
            element: JsonElement,
            type: Type,
            ctx: JsonDeserializationContext
        ): ChatEmotesConfig {
            val root = element.asJsonObject
            if (!GsonHelper.isStringValue(root, "version")) {
                throw JsonSyntaxException(
                    "Expected version to be a string, got " + GsonHelper.getType(root["version"])
                )
            }
            val version: Version = try {
                Version.parse(GsonHelper.getAsString(root, "version"))
            } catch (ex: VersionParsingException) {
                throw JsonSyntaxException("Invalid version", ex)
            }
            if (version.compareTo(modMetadata!!.version) != 0) {
                throw JsonParseException("Unsupported version of config: " + version.friendlyString)
            }
            val resourcepacks = ctx.deserialize<List<ResourcepackConfig>>(
                GsonHelper.getAsJsonObject(root, "resourcepacks"),
                ResourcepackConfig.LIST_TYPE
            )
            return ChatEmotesConfig(resourcepacks)
        }
    }

    companion object {
        private val GSON = GsonBuilder()
            .registerTypeAdapter(ResourcepackConfig.LIST_TYPE, ListSerializer)
            .registerTypeAdapter(ChatEmotesConfig::class.java, Serializer)
            .create()

        @Throws(IOException::class, JsonParseException::class)
        fun load(input: InputStream?): ChatEmotesConfig {
            create(input).use { reader -> return GSON.fromJson(reader, ChatEmotesConfig::class.java) }
        }
    }
}